src/V3Timing.cpp

```diff
@@ -0,0 +1,628 @@
+        // Create the intermediate select vars
+        nodep->lhsp()->foreach<AstSel>([&](AstSel* selp) {
+            if (VN_IS(selp->lsbp(), Const)) return;
+            AstVarScope* const lsbVscp
+                = m_scopep->createTemp(m_intraLsbNames.get(nodep), selp->lsbp()->dtypep());
+            controlp->addHereThisAsNext(new AstAssign{
+                flp, new AstVarRef{flp, lsbVscp, VAccess::WRITE}, selp->lsbp()->unlinkFrBack()});
+            selp->lsbp(new AstVarRef{flp, lsbVscp, VAccess::READ});
+            // widthp should be const
+        });
+        nodep->lhsp()->foreach<AstNodeSel>([&](AstNodeSel* selp) {
+            if (VN_IS(selp->bitp(), Const)) return;
+            AstVarScope* const bitVscp
+                = m_scopep->createTemp(m_intraIndexNames.get(nodep), selp->bitp()->dtypep());
+            controlp->addHereThisAsNext(new AstAssign{
+                flp, new AstVarRef{flp, bitVscp, VAccess::WRITE}, selp->bitp()->unlinkFrBack()});
+            selp->bitp(new AstVarRef{flp, bitVscp, VAccess::READ});
+        });
+        // Create the intermediate value var
+        AstVarScope* const newvscp
+            = m_scopep->createTemp(m_intraValueNames.get(nodep), nodep->lhsp()->dtypep());
+        // Assign the RHS to the intermediate before the timing control
+        auto* const assignp = new AstAssign{flp, new AstVarRef{flp, newvscp, VAccess::WRITE},
+                                            nodep->rhsp()->unlinkFrBack()};
+        // Assign the intermediate to the LHS after the timing control
+        nodep->rhsp(new AstVarRef{flp, newvscp, VAccess::READ});
+        if (VN_IS(nodep, AssignDly)) {
+            // If it's an NBA with an intra assignment delay, put it in a fork
+            controlp->replaceWith(assignp);
+            auto* const forkp = new AstFork{flp, "", controlp};
+            forkp->joinType(VJoinType::JOIN_NONE);
+            assignp->addNextHere(forkp);
+        } else {
+            controlp->addHereThisAsNext(assignp);
+        }
+    }
+    virtual void visit(AstAssignW* nodep) override {
+        iterateChildren(nodep);
+        if (!nodep->timingControlp()) return;
+        FileLine* const flp = nodep->fileline();
+        // Create a sentree from the RHS, we'll need that for a new always
+        AstSenItem* const senItemsp = varRefpsToSenItemsp(nodep->rhsp());
+        // Convert it to an always; the new assign with intra delay will be handled by
+        // visit(AstNodeAssign*)
+        auto* const alwaysp = nodep->convertToAlways();
+        alwaysp->setSuspendable();
+        // Do not move this always into the initial domain
+        alwaysp->user1(true);
+        if (senItemsp) {
+            auto* const activep = new AstActive{flp, "", new AstSenTree{flp, senItemsp}};
+            activep->sensesStorep(activep->sensesp());
+            activep->addStmtsp(alwaysp->unlinkFrBack());
+            m_activep->addNextHere(activep);
+        }
```

Comment:
1. It can be combo, but then for code like this:
   ```systemverilog
   assign #1 clk = ~clk;
   ```
   we get a cycle (and `UNOPTFLAT`). I could `modifyWarnOff` on all variables written in the assign.
2. You're correct, it's more like a fork. I can put the assignment in a fork, it shouldn't cause problems.

Reply:
> It can be combo, but then for code like this:
assign #1 clk = ~clk;
we get a cycle (and UNOPTFLAT). I could modifyWarnOff on all variables written in the assign.

Gah, ok I see it. But still, only doing these self referentials (which are presumably rare?) as clocked and leaving all else comb might help with speed at some point

> You're correct, it's more like a fork. I can put the assignment in a fork, it shouldn't cause problems.

My point was about using `alwaysp->setSuspendable();` and if that is correct. If you use a fork *instead* of the always, and that yields the same code/behavour, then I think that is easier to understand. If you need the fork *and* the always, and the fork does not add much, then don't worry, maybe explain in a comment.

Reply:
Resolved the `UNOPTFLAT` by adding an intermediate variable on the LHS of the `AssignW` and disabling the warning on it. A better solution would involve teaching `V3SchedAcyclic` to handle awaits, but I'm not sure how to do that.

Reply:
Makes sense. It'll be interesting to revisit this whole thing with dataflow analysis in our toolbox.